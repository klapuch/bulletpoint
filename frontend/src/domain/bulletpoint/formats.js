import reactStringReplace from 'react-string-replace';
import { Link } from 'react-router-dom';
import getSlug from 'speakingurl';
import React from 'react';
import type { FetchedBulletpointType } from './types';

const REGEX = /\[\[(.+?)\]\]/g;

export const numberOfReferences = (text: string) => (text.match(REGEX) || []).length;

export const replaceMatches = (bulletpoint: FetchedBulletpointType) => {
  const { referenced_theme: referencedTheme } = bulletpoint;
  let order = 0;
  return reactStringReplace(bulletpoint.content, REGEX, (match) => {
    const currentTheme = referencedTheme[order];
    order += 1;
    return (
      <Link key={order} to={`/themes/${currentTheme.id}/${getSlug(currentTheme.name)}`}>
        {match}
      </Link>
    );
  });
};

export const withComparisons = (content: string, bulletpoint: FetchedBulletpointType) => {
  const { compared_theme: comparedTheme } = bulletpoint;
  if (comparedTheme.length === 0) {
    return content;
  }
  return (
    <em>
      {content}
      {' '}
&hellip; neÅ¾&nbsp;
      {comparedTheme.map((theme, order) => (
        <React.Fragment key={order}>
          {order === 0 ? '' : ', '}
          <Link to={`/themes/${theme.id}/${getSlug(theme.name)}`}>
            {theme.name}
          </Link>
        </React.Fragment>
      ))}
    </em>
  );
};
