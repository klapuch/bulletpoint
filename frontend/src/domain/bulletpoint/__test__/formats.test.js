import { withComparisons } from '../formats';

test('adding than part', () => {
  expect(
    withComparisons(
      'This is better',
      {
        compared_theme: [
          { id: 1, name: 'everything' },
        ],
      },
    ),
  ).toMatchSnapshot();

  expect(
    withComparisons(
      'This is better',
      {
        compared_theme: [
          { id: 1, name: 'everything' },
          { id: 2, name: 'else' },
        ],
      },
    ),
  ).toMatchSnapshot();
});

test('without than part for empty', () => {
  expect(
    withComparisons(
      'This good enough',
      { compared_theme: [] },
    ),
  ).toMatchSnapshot();
});
