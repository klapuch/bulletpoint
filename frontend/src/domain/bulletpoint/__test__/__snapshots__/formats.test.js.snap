// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adding than part 1`] = `
<em>
  This is better
   
  … než 
  <React.Fragment>
    
    <Link
      to="/themes/1/everything"
    >
      everything
    </Link>
  </React.Fragment>
</em>
`;

exports[`adding than part 2`] = `
<em>
  This is better
   
  … než 
  <React.Fragment>
    
    <Link
      to="/themes/1/everything"
    >
      everything
    </Link>
  </React.Fragment>
  <React.Fragment>
    , 
    <Link
      to="/themes/2/else"
    >
      else
    </Link>
  </React.Fragment>
</em>
`;

exports[`replacing by positions 1`] = `
Array [
  "",
  <Link
    to="/themes/1/php"
  >
    PHP
  </Link>,
  " is ",
  <Link
    to="/themes/2/acronym"
  >
    acronym
  </Link>,
  "",
]
`;

exports[`text as nothing to replae 1`] = `
Array [
  "PHP is acronym",
]
`;

exports[`without than part for empty 1`] = `"This good enough"`;
